rfc/0019-reflectors-database-cache.md

```diff
@@ -0,0 +1,164 @@
+For this reason, the reflectors will be created at boot time instead of being created lazily.
+
+The new flow will be as follows:
+
+1. The policy server starts.
+2. The policy server creates the database and the connection pool.
+3. The policy server iterates over the resource kinds that are being set in the policies configuration and creates the tables and the reflectors.
+4. The reflector receive a `watcher::Event::Restarted` event and fetches the objects from the Kubernetes API server and stores them in the database.
+5. The reflector keeps the objects up-to-date by watching for changes in the Kubernetes API server, receiving the events `watcher::Event::Added`, `watcher::Event::Modified` and `watcher::Event::Deleted`.
+6. When a policy requests a resource or a list of resources, the policy server queries the database returns the objects to the policy callback channel.
+
+## Querying the database
+
+The host callback handler will be modified to query the database and return the objects to the policy callback channel.
+
+Get the pod named `nginx` in the `default` namespace:
+
+```sql
+SELECT object FROM v1_pods WHERE name = 'nginx' AND namespace = 'default';
+```
+
+Get all the pods in the `default` namespace:
+
+```sql
+SELECT object FROM v1_pods WHERE namespace = 'default';
+```
+
+Get all the pods with the label `app=nginx`:
+
+```sql
+SELECT object from v1_pods where json_extract(object, '$.metadata.labels.app') = 'nginx';
+```
+
+## Gatekeeper inventory cache
+
+Gatekeeper policies require a different approach, as they expect the cache data to be present under the `data.inventory` document.
+To know more about the Gatekeeper inventory document, please refer to the [Gatekeeper documentation](https://open-policy-agent.github.io/gatekeeper/website/docs/sync/#accessing-replicated-data).
+
+At the time of writing, we store the serialized inventory in memory to speed up the evaluation of Gatekeeper policies.
+Even though this solves the problem of having to wait for the inventory to be built and serialized at policy evaluation time from the resource-based reflectors,
+it can lead to high memory usage in the policy server specially in large clusters.
+
+This RFC proposes to create a dedicated table in the database to store the inventory cache.
+The inventory is built at boot time and kept up-to-date by watching for changes in the Kubernetes API server.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+At boot time the policy server will start the reflectors.
+Once a reflector is started it will receive a `watcher::Event::Restarted` event with all the objects of the resource kind it is watching.
+Even if the request is paginated (see: https://docs.rs/kube/latest/kube/runtime/watcher/struct.Config.html#method.page_size)
+to reduce Kubernetes API and serialization load, the items will be buffered-up in the watcher event causing a spike in memory usage.
```

Comment:
As far as I can see, we will not wait for the initial load of the database. Therefore, I think that another drawback is that right after the starting of the policy server, if it receives a evaluation request that requires many resources that are still being loaded,  the policy can evaluate using incomplete data. Because the reflectors will be still loading the data into the database. Leading to invalid resources being accepted in the cluster. 

I don't believe that will be a huge problem now. Because we also have the audit scanner to inspect the cluster. However, I believe this can be important for users that have policies that requires access to massive resources count. Thus,  I think is important to document that.